fdbserver/ptxn/MessageSerializer.cpp

```diff
@@ -0,0 +1,347 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbserver/ptxn/MessageSerializer.h"
+
+#include <string>
+
+#include "fdbclient/FDBTypes.h"
+#include "flow/Error.h"
+#include "flow/String.h"
+
+namespace ptxn {
+
+SubsequencedMessageSerializer::SubsequencedMessageSerializer(const StorageTeamID& storageTeamID) {
+	header.storageTeamID = storageTeamID;
+	header.lastVersion = invalidVersion;
+	header.lastSubsequence = invalidSubsequence;
+}
+
+void SubsequencedMessageSerializer::startVersionWriting(const Version& version) {
+	// Ensure the ordering of version.
+	ASSERT(version > header.lastVersion);
+
+	if (header.firstVersion == invalidVersion) {
+		header.firstVersion = version;
+	}
+	header.lastVersion = version;
+	header.lastSubsequence = invalidSubsequence;
+
+	serializer.startNewSection();
+}
+
+void SubsequencedMessageSerializer::write(const SubsequenceMutationItem& subsequenceMutationItem) {
+	ASSERT(subsequenceMutationItem.subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequenceMutationItem.subsequence;
+	serializer.writeItem(subsequenceMutationItem);
+}
+
+void SubsequencedMessageSerializer::write(const Subsequence& subsequence, const MutationRef& mutation) {
+	ASSERT(subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequence;
+	serializer.writeItem(SubsequenceMutationItem{ subsequence, mutation });
+}
+
+void SubsequencedMessageSerializer::write(const SubsequenceSpanContextItem& subsequenceSpanContextItem) {
+	ASSERT(subsequenceSpanContextItem.subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequenceSpanContextItem.subsequence;
+	serializer.writeItem(subsequenceSpanContextItem);
+}
+
+void SubsequencedMessageSerializer::write(const Subsequence& subsequence, const SpanContextMessage& spanContext) {
+	ASSERT(subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequence;
+	serializer.writeItem(SubsequenceSpanContextItem{ subsequence, spanContext });
+}
+
+void SubsequencedMessageSerializer::write(const SubsequenceLogProtocolMessageItem& subsequenceLogProtocolMessageItem) {
+	ASSERT(subsequenceLogProtocolMessageItem.subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequenceLogProtocolMessageItem.subsequence;
+	serializer.writeItem(subsequenceLogProtocolMessageItem);
+}
+
+void SubsequencedMessageSerializer::write(const Subsequence& subsequence,
+                                          const LogProtocolMessage& logProtocolMessage) {
+	ASSERT(subsequence > header.lastSubsequence);
+
+	header.lastSubsequence = subsequence;
+	serializer.writeItem(SubsequenceLogProtocolMessageItem{ subsequence, logProtocolMessage });
+}
+
+void SubsequencedMessageSerializer::write(
+    const Subsequence& subsequence,
+    const std::variant<MutationRef, SpanContextMessage, LogProtocolMessage>& message) {
+
+	switch (static_cast<MutationRef::Type>(message.index())) {
```

Comment:
what if `message.index()` returns `variant_npos`?

Reply:
Yeah, actually this function is not written properly. Let me revisit this together with corresponding tests.

Reply:
Yeah, actually this function is not written properly. Let me revisit this together with corresponding tests.